src/ripple/app/misc/SHAMapStoreImp.cpp

```diff
@@ -664,11 +664,11 @@ SHAMapStoreImp::clearPrior (LedgerIndex lastRotated)
                 << deleteBatch << " rows.";
             long long totalRowsAffected = 0;
             long long rowsAffected;
-            soci::statement st = [&]
+            boost::optional<soci::statement> st;
             {
                 auto db = ledgerDb_->checkoutDb();
-                return (db->prepare << deleteQuery);
-            }();
+                st.emplace(db->prepare << deleteQuery);
```

Comment:
`statement` is a handle class. Constructing the `statement_impl` needs to happen while holding the db lock. This happens in the `statement(details::prepare_temp_type const & prep)` constructor. Coping a `statement` does not need to happen in the lock. The copy constructor just copies the pointer and increments the ref count. The `session` is never touched.

The old code was broken because lambda returned a `prepare_temp_type` and constructed a `statement_impl` without holding the lock. Changing the lambda to return a `statement` fixes this, and that solution seems less exotic than using `boost::optional`.

Proposed:
```
            auto st = [&]
            {
                auto db = ledgerDb_->checkoutDb();
                return soci::statement (db->prepare << deleteQuery);
            }();
```
(This works both with and without RVO as well).